src/compiler/checker.ts

```diff
@@ -19549,7 +19563,7 @@ namespace ts {
                 if (sourceFlags & TypeFlags.TypeVariable) {
                     // IndexedAccess comparisons are handled above in the `targetFlags & TypeFlage.IndexedAccess` branch
                     if (!(sourceFlags & TypeFlags.IndexedAccess && targetFlags & TypeFlags.IndexedAccess)) {
-                        const constraint = getConstraintOfType(source as TypeVariable);
+                        const constraint = getConstraintOfType(source as TypeVariable) || unknownType;
                         if (!strictNullChecks && (!constraint || (sourceFlags & TypeFlags.TypeParameter && constraint.flags & TypeFlags.Any))) {

```

Comment:
Since now `constraint` is no longer nullable this can be simplified:
```suggestion
                        if (!strictNullChecks && (sourceFlags & TypeFlags.TypeParameter && constraint.flags & TypeFlags.Any)) {

```

Reply:
Good catch for the wrong reason - this rule's only here for compat in non-strict mode; as such, it probably still needs to check for constraint presence (as inconsistent as that is). Otherwise it should also look for `TypeFlags.Unknown` on the constraint in addition to `Any` (which'd still change the assignability of `T extends unknown` in non-strict mode).